pkg/validation/components_test.go

```diff
@@ -532,14 +546,19 @@ func TestValidateComponents(t *testing.T) {
 		t.Run(tt.name, func(t *testing.T) {
 			err := ValidateComponents(tt.components)
 
-			if merr, ok := err.(*multierror.Error); ok && tt.wantErr != nil {
-				if assert.Equal(t, len(tt.wantErr), len(merr.Errors), "Error list length should match") {
-					for i := 0; i < len(merr.Errors); i++ {
-						assert.Regexp(t, tt.wantErr[i], merr.Errors[i].Error(), "Error message should match")
+			merr, ok := err.(*multierror.Error)
+			if ok {
+				if tt.wantErr != nil {
+					if assert.Equal(t, len(tt.wantErr), len(merr.Errors), "Error list length should match") {
+						for i := 0; i < len(merr.Errors); i++ {
+							assert.Regexp(t, tt.wantErr[i], merr.Errors[i].Error(), "Error message should match")
+						}
 					}
+				} else {
+					assert.Equal(t, nil, err, "Error should be nil")
 				}
```

Comment:
this else branch is for line 551 -> `if tt.wantErr!= nil`, but `ok==true`. which means we want the error to be nil, but we got something returned from the `ValidateComponents`

Reply:
I have updated to use `t.Errorf` instead of `assert.Equal` to avoid confusion.

Reply:
I think the issue is that line 551 is never reached, though:
```go
merr, ok := err.(*multierror.Error) // If err is nil, then we'll have merr == nil, ok == false
if ok {                             // ok == true implies err != nil
  if tt.wantErr != nil {
    if assert.Equal(t, len(tt.wantErr), len(merr.Errors), "Error list length should match") {
      for i := 0; i < len(merr.Errors); i++ {
        assert.Regexp(t, tt.wantErr[i], merr.Errors[i].Error(), "Error message should match")
      }
    }
  } else {
    assert.Equal(t, nil, err, "Error should be nil")
  }
```

Reply:
```
merr, ok := err.(*multierror.Error) 
if ok {                           // err != nil
  if tt.wantErr != nil { // err != nil && wantErr != nil, need to process and check if each error matches
    if assert.Equal(t, len(tt.wantErr), len(merr.Errors), "Error list length should match") {
      for i := 0; i < len(merr.Errors); i++ {
        assert.Regexp(t, tt.wantErr[i], merr.Errors[i].Error(), "Error message should match")
      }
    }
  } else { // this is a else statement of if wantErr != nil. so err != nil && wantErr == nil. and test to error out
    t.Errorf("Error should be nil, got %v", err)
  }
```

Reply:
You're right -- maybe my brain just isn't working today :slightly_smiling_face:. Sorry for the hassle.

To try and explain where I'm coming from -- I feel like it'd be simpler to check whether an error exists first before asserting the type. As it stands, the case "returns nil and we expect nil" is not explicitly checked, and is verified only by dodging two conditionals.

| | Returns `nil` | Returns `multierror.Error` | Returns other error |
| -- | -- | -- | -- |
| **Expect no error** | Not checked directly | Checked line 558 | Test succeeds even though an error was returned | 
| **Expect an error**| Checked at line 561| Checked lines 551-556 | Test fails but prints "Error should not be nil" even though error is not nil |

I understand (after reading the validation code) that only multierror.Error is returned, but from reading the test case, this behavior is not clear, as there's no explicit assertion that _any_ error returned from `ValidateComponents` has to be of type multierror.Error.

Either way, I'll approve PR so it can be merged.